src/algebra/gcd_domain.lean

```diff
@@ -27,41 +27,108 @@ attribute [simp] norm_unit_coe_units norm_unit_zero norm_unit_mul
 section normalization_domain
 variable [normalization_domain α]
 
+def normalize (x : α) : α :=
+x * norm_unit x
+
+theorem associated_normalize {x : α} : associated x (normalize x) :=
+⟨_, rfl⟩
+
+theorem normalize_associated {x : α} : associated (normalize x) x :=
+associated_normalize.symm
+
 @[simp] theorem norm_unit_one : norm_unit (1:α) = 1 :=
 norm_unit_coe_units 1
 
+variables (α)
+@[simp] lemma normalize_zero : normalize (0 : α) = 0 :=
+by rw [normalize, zero_mul]
+
+@[simp] lemma normalize_one : normalize (1 : α) = 1 :=
+by rw [normalize, norm_unit_one, units.coe_one, mul_one]
+variables {α}
+
+lemma normalize_coe_units (u : units α) : normalize (u : α) = 1 :=
+by rw [normalize, norm_unit_coe_units, ← units.coe_mul, mul_inv_self, units.coe_one]
+
+theorem normalize_mul (x y : α) : normalize (x * y) = normalize x * normalize y :=
+classical.by_cases (λ hx : x = 0, by rw [hx, zero_mul, normalize_zero, zero_mul]) $ λ hx,
+classical.by_cases (λ hy : y = 0, by rw [hy, mul_zero, normalize_zero, mul_zero]) $ λ hy,
+by simp only [normalize, norm_unit_mul hx hy, units.coe_mul]; simp only [mul_assoc, mul_left_comm y]
+
+@[simp] lemma normalize_eq_zero {x : α} : normalize x = 0 ↔ x = 0 :=
+⟨λ hx, (associated_zero_iff_eq_zero x).1 $ hx ▸ associated_normalize, by rintro rfl; exact normalize_zero α⟩
+
+@[simp] lemma normalize_eq_one {x : α} : normalize x = 1 ↔ is_unit x :=
+⟨λ hx, is_unit_iff_exists_inv.2 ⟨_, hx⟩, λ ⟨u, hu⟩, hu.symm ▸ normalize_coe_units u⟩
+
```

Comment:
debatable whether this should be `simp`. Equalities are good.

Reply:
done